Realm/RLMManagedDictionary.mm

```diff
@@ -0,0 +1,688 @@
+    realm::object_store::Dictionary _backingCollection;
+}
+@end
+
+@interface RLMDictionaryChange ()
+- (instancetype)initWithChanges:(realm::DictionaryChangeSet)changes previous:(realm::Dictionary&)collection;
+@end
+
+namespace {
+struct DictionaryCallbackWrapper {
+    void (^block)(id, RLMDictionaryChange *, NSError *);
+    RLMManagedDictionary *collection;
+    realm::TransactionRef previousTransaction;
+    std::shared_ptr<realm::Dictionary> previousDictionary;
+
+    DictionaryCallbackWrapper(void (^block)(id, RLMDictionaryChange *, NSError *), RLMManagedDictionary *dictionary)
+    : block(block)
+    , collection(dictionary)
+    {
+        previousTransaction = static_cast<realm::Transaction&>(collection.realm.group).duplicate();
+        previousDictionary.reset(static_cast<realm::Dictionary*>(previousTransaction->import_copy_of(collection->_backingCollection.get_impl()).release()));
+
+    }
+
+    void operator()(realm::DictionaryChangeSet const& changes, std::exception_ptr err) {
+        if (err) {
+            previousTransaction->end_read();
+            try {
+                rethrow_exception(err);
+            }
+            catch (...) {
+                NSError *error = nil;
+                RLMRealmTranslateException(&error);
+                block(nil, nil, error);
+                return;
+            }
+        }
+
+        if (changes.deletions.empty() &&
+            changes.insertions.empty() &&
+            changes.modifications.empty()) {
+            block(collection, nil, nil);
+        }
+        else {
+            block(collection, [[RLMDictionaryChange alloc] initWithChanges:changes previous:*previousDictionary], nil);
+        }
+        if (collection.isInvalidated) {
+            previousTransaction->end_read();
+        }
+        else {
+            previousTransaction->advance_read(static_cast<realm::Transaction&>(collection.realm.group).get_version_of_current_transaction());
+        }
+    }
+};
+} //anonymous namespace
+
+@implementation RLMDictionaryChange {
+    realm::DictionaryChangeSet _changes;
+    NSArray<id> *_deletions;
+}
+
+- (instancetype)initWithChanges:(realm::DictionaryChangeSet)changes previous:(realm::Dictionary&)dictionary {
+    self = [super init];
+    if (self) {
+        _changes = std::move(changes);
+        if (_changes.deletions.empty()) {
+            _deletions = @[];
+        }
+        else {
+            NSMutableArray *array = [NSMutableArray arrayWithCapacity:_changes.deletions.size()];
+            for (auto index : _changes.deletions) {
+                realm::Mixed key = dictionary.get_key(index);
+                switch (key.get_type()) {
+                    case realm::type_String:
+                        [array addObject:@(key.get_string().data())];
+                        break;
+                    default:
+                        // Don't throw so older SDK versions can handle any new key types.
+                        break;
+                }
```

Comment:
Using RLMMixedToObjc would be both simpler and more forward-compatible.

Reply:
My rational was they it would add more overhead